client/client.go

```diff
@@ -165,3 +166,44 @@ func (c *Client) Pay(channelId types.Destination, amount *big.Int) {
 	// Send the event to the engine
 	c.engine.PaymentRequestsFromAPI <- engine.PaymentRequest{ChannelId: channelId, Amount: amount}
 }
+
+// WaitForObjectivesToComplete returns a channel on which
+// a mapping showing which of the supplied objective ids is still incomplete
+// will be sent either when all objectives have been completed, or when the supplied context is cancelled.
+func (c *Client) WaitForObjectivesToComplete(ctx context.Context, ids ...protocols.ObjectiveId) chan map[protocols.ObjectiveId]bool {
+
+	doneChan := make(chan map[protocols.ObjectiveId]bool)
+
+	incomplete := make(map[protocols.ObjectiveId]bool)
+	for _, id := range ids {
+		incomplete[id] = true
+	}
+
+	handleCompletedObjectiveAndCheckAllCompleted := func(got protocols.ObjectiveId) bool {
+		allCompleted := true
+		for _, id := range ids {
+			if id == got {
+				incomplete[id] = false
+			} // TODO if the id is irrelevant to us, we could resend it on the CompletedObjective chan for others to read?
+			allCompleted = allCompleted && !incomplete[id]
+		}
+		return allCompleted
+	}
+
+	go func() {
+		for {
+			select {
+			case got := <-c.CompletedObjectives():
```

Comment:
I think this means that once `WaitForObjectivesToComplete` is called it will consume all completed objectives from `c.CompletedObjectives()`, even if the objectives are unrelated to the `ids` we passed in. This might be confusing for clients, ie if they call `WaitForObjectivesToComplete` for a long running objective it will appear like no other objectives complete while they're waiting.

 There could also be contention on the channel between `WaitForObjectivesToComplete` and some client code pulling from `c.CompletedObjectives()`?

Reply:
Agreed. Did you come up with a solution to these problems in your completion manager code?

Perhaps `WaitForObjectivesToComplete` could register an "event handler" (i.e. create a new `chan`) scoped to a specific objective id. The engine would then send on each registered event handler when that objective completes, which would eliminate contention and prevent consuming unrelated events.